pkg/login/social/connectors/azuread_oauth.go

```diff
@@ -168,6 +170,103 @@ func (s *SocialAzureAD) UserInfo(ctx context.Context, client *http.Client, token
 	return userInfo, nil
 }
 
+func (s *SocialAzureAD) Exchange(ctx context.Context, code string, authOptions ...oauth2.AuthCodeOption) (*oauth2.Token, error) {
+	s.reloadMutex.RLock()
+	defer s.reloadMutex.RUnlock()
+
+	oauthCfg := s.GetOAuthInfo()
+
+	// Handle client authentication types
+	switch oauthCfg.ClientAuthentication {
+	case social.ClientSecretJWT:
+		var clientAssertion string
+		var err error
+
+		// Generate client assertion based on ManagedIdentityClientID
+		if oauthCfg.ManagedIdentityClientID != "" {
+			clientAssertion, err = s.ManagedIdentityCallback(ctx)
+		} else {
+			clientAssertion, err = createJWT(*oauthCfg)
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		authOptions = append(authOptions,
+			oauth2.SetAuthURLParam("client_assertion", clientAssertion),
+			oauth2.SetAuthURLParam("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer"),
+		)
+
+		// Remove client_secret to avoid sending it in the request
+		s.Config.ClientSecret = ""
+
+	case social.ClientSecretPost:
+		// Use default behavior for ClientSecretPost, no additional setup needed
+
+	default:
+		return nil, fmt.Errorf("invalid client authentication method: %s", oauthCfg.ClientAuthentication)
+	}
+
+	// Default exchange method
+	return s.Config.Exchange(ctx, code, authOptions...)
+}
+
+// Creates and signs a JWT token using the shared client secret. Based on OIDC specification for client_secret_jwt.
+func createJWT(oauthCfg social.OAuthInfo) (string, error) {
+	now := time.Now().UTC()
+	claims := jwt.Claims{
+		Issuer:    oauthCfg.ClientId,
+		Subject:   oauthCfg.ClientId,
+		Audience:  jwt.Audience{oauthCfg.TokenUrl},
+		ID:        uuid.New().String(),
+		Expiry:    jwt.NewNumericDate(now.Add(5 * time.Minute)),
+		NotBefore: jwt.NewNumericDate(now),
+		IssuedAt:  jwt.NewNumericDate(now),
+	}
+
+	sharedSecret := oauthCfg.ClientSecret
+	signer, err := jose.NewSigner(jose.SigningKey{Algorithm: jose.HS256, Key: []byte(sharedSecret)}, nil)
+	if err != nil {
+		return "", err
+	}
+	jwt, err := jwt.Signed(signer).Claims(claims).CompactSerialize()
+	if err != nil {
+		return "", err
+	}
+
+	return jwt, nil
+}
+
+// ManagedIdentityCallback retrieves a token using the managed identity credential of the Azure service.
+func (s *SocialAzureAD) ManagedIdentityCallback(ctx context.Context) (string, error) {
+	// exchange auth code to a valid token
+	managedIdentityClientID := s.GetOAuthInfo().ManagedIdentityClientID
+	azScopes := []string{"api://AzureADTokenExchange/.default"}
```

Comment:
@kalleep We will also need a flag in the .ini file for the audience, since this is changes based on which cloud environment the Grafana instance is deployed in. For now, I will label this new flag as `audience`. Let me know if you have another idea for the naming